image-mdraid.c

```diff
@@ -0,0 +1,226 @@
+	sb->ctime=time(NULL);	/* lo 40 bits are seconds, top 24 are microseconds or 0*/
+
+	sb->level=1;		/* -4 (multipath), -1 (linear), 0,1,4,5 */
+	//sb->layout=2;		/* only for raid5 and raid10 currently */
+	sb->size=(image->size - DATA_OFFSET_BYTES)/512;	/* used size of component devices, in 512byte sectors */
+
+	sb->chunksize=0;		/* in 512byte sectors - not used in raid 1 */
+	sb->raid_disks=1;
+	sb->bitmap_offset=8;	/* sectors after start of superblock that bitmap starts
+					 * NOTE: signed, so bitmap can be before superblock
+					 * only meaningful of feature_map[0] is set.
+					 */
+
+	/* constant this-device information - 64 bytes */
+	sb->data_offset=DATA_OFFSET_SECTORS;	/* sector start of data, often 0 */
+	sb->data_size=(image->size - DATA_OFFSET_BYTES)/512;	/* sectors in this device that can be used for data */
+	sb->super_offset=8;	/* sector start of this superblock */
+
+	sb->dev_number=0;	/* permanent identifier of this  device - not role in raid */
+	sb->cnt_corrected_read=0; /* number of read errors that were corrected by re-writing */
+	random_uuid(sb->device_uuid); /* user-space setable, ignored by kernel U8[16] */ //TODO: should we allow user to set this?
+	sb->devflags=0;	/* per-device flags.  Only two defined...*/
+		//#define	WriteMostly1	1	/* mask for writemostly flag in above */
+		//#define	FailFast1	2	/* Should avoid retries and fixups and just fail */
+
+		/* Bad block log.  If there are any bad blocks the feature flag is set.
+		* If offset and size are non-zero, that space is reserved and available
+		*/
+	sb->bblog_shift=9;    /* shift from sectors to block size */ //TODO: not sure why this is 9
+	sb->bblog_size=8; /* number of sectors reserved for list */
+	sb->bblog_offset=16;   /* sector offset from superblock to bblog,
+			* signed - not unsigned */
+
+	/* array state information - 64 bytes */
+	sb->utime=0;		/* 40 bits second, 24 bits microseconds */
+	sb->events=0;		/* incremented when superblock updated */
+	sb->resync_offset=0;	/* data before this offset (from data_offset) known to be in sync */
+	sb->max_dev=max_devices; /* size of devs[] array to consider */
+	//__u8	pad3[64-32];	/* set to 0 when writing */
+
+	/* device state information. Indexed by dev_number.
+	 * 2 bytes per device
+	 * Note there are no per-device state flags. State information is rolled
+	 * into the 'roles' value.  If a device is spare or faulty, then it doesn't
+	 * have a meaningful role.
+	 */
+	//__le16	dev_roles[];	/* role in array, or 0xffff for a spare, or 0xfffe for faulty */
+
+
+	//Calculate checksum
+	sb->sb_csum=calc_sb_1_csum(sb);
+
+
+	//construct image file
+	int ret;
+	ret = prepare_image(image, image->size);
+	if(ret) return ret;
+	ret = insert_data(image, sb, imageoutfile(image), sizeof(struct mdp_superblock_1)+max_devices*2, 8*512);
+	if(ret) return ret;
+	if(img_in) {
+		ret = insert_image(image, img_in, image->size-DATA_OFFSET_BYTES, DATA_OFFSET_BYTES, 0);
```

Comment:
Why not img_in->size for the size argument?

Reply:
because img_in might be smaller than raid itself... i think it's valid use case. eg.: i download small premade-image from internet, but i know i want my partition to be bigger.